deprecated zmq send/recv calls

Seems as of ZMQ 4.3.1 some of these calls have become deprecated.
```
[140/228] Building CXX object src/zeromq/CMakeFiles/goby_zeromq.dir/transport/interprocess.cpp.o
../src/zeromq/transport/interprocess.cpp: In member function 'bool goby::zeromq::InterProcessPortalMainThread::recv(goby::zeromq::protobuf::InprocControl*, int)':
../src/zeromq/transport/interprocess.cpp:74:45: warning: 'bool zmq::detail::socket_base::recv(zmq::message_t*, int)' is deprecated: from 4.3.1, use recv taking a reference to message_t and recv_flags [-Wdeprecated-declarations]
     if (control_socket_.recv(&zmq_msg, flags))
                                             ^
In file included from ../src/zeromq/transport/interprocess.h:27,
                 from ../src/zeromq/transport/interprocess.cpp:25:
/usr/include/zmq.hpp:1267:10: note: declared here
     bool recv(message_t *msg_, int flags_ = 0)
          ^~~~
../src/zeromq/transport/interprocess.cpp: In member function 'void goby::zeromq::InterProcessPortalMainThread::publish(const string&, const char*, int)':
../src/zeromq/transport/interprocess.cpp:107:33: warning: 'bool zmq::detail::socket_base::send(zmq::message_t&, int)' is deprecated: from 4.3.1, use send taking message_t and send_flags [-Wdeprecated-declarations]
         publish_socket_.send(msg);
                                 ^
In file included from ../src/zeromq/transport/interprocess.h:27,
                 from ../src/zeromq/transport/interprocess.cpp:25:
/usr/include/zmq.hpp:1191:10: note: declared here
     bool send(message_t &msg_,
          ^~~~
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -410,6 +410,9 @@ jobs:
     docker:
       - image: ubuntu:focal
     working_directory: /root/goby3
+    environment:
+      DEBIAN_FRONTEND: "noninteractive"
+      DEBIAN_PRIORITY: "critical"
     steps:
       - run:
           name: Install packages server

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -178,17 +178,18 @@ job-template-deb-amd64: &job-template-deb-amd64
 job-template-deb-cross: &job-template-deb-cross
   <<: *job-template-deb-amd64
 
-# base sanitizer and upload off Bionic build
+# base sanitizer off Focal build
 job-template-sanitizers: &job-template-sanitizers
   <<: *job-template-amd64
   environment:
     <<: *environment-template-common
-    <<: *environment-template-bionic
+    <<: *environment-template-focal
     <<: *environment-template-amd64
     SANITIZER_NUM_JOBS: 2
     TSAN_OPTIONS: "suppressions=/root/goby3/src/test/suppressions.tsan.txt"    
-  docker: *docker-base-bionic
+  docker: *docker-base-focal
 
+# Base upload off Bionic build
 job-template-upload: &job-template-upload
   <<: *job-template-amd64
   docker: *docker-base-bionic
@@ -219,8 +220,9 @@ workflows:
           <<: *filter-template-non-master
       - amd64-buster-build:
           <<: *filter-template-non-master
-      - amd64-xenial-build:
-          <<: *filter-template-non-master
+# To keep basic builds to a minimum, remove Xenial
+#      - amd64-xenial-build:
+#          <<: *filter-template-non-master
       - amd64-bionic-build:
           <<: *filter-template-non-master
       - amd64-focal-build:

--- a/.circleci/docker/focal/arm64/Dockerfile
+++ b/.circleci/docker/focal/arm64/Dockerfile
@@ -1,0 +1,16 @@
+FROM gobysoft/goby3-ubuntu-build-base:20.04.1
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+# Overwrite non-multiarch packages
+RUN apt-get update && \
+    apt-get -y install libdccl3-dev:arm64 \
+            libcrypto++-dev:arm64 \
+            libcrypto++6:arm64 \
+            libwt-dev:arm64 libwtdbo-dev:arm64 libwtdbosqlite-dev:arm64 libwthttp-dev:arm64 \
+            libboost-regex-dev:arm64 libicu-dev:arm64 \
+            libhdf5-dev:arm64 \
+            libgmp3-dev:arm64 \
+            libstdc++-7-dev:arm64 \
+            && \
+    rm -rf /var/lib/apt/lists/*

--- a/.circleci/docker/focal/armhf/Dockerfile
+++ b/.circleci/docker/focal/armhf/Dockerfile
@@ -1,0 +1,16 @@
+FROM gobysoft/goby3-ubuntu-build-base:20.04.1
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+# Overwrite non-multiarch packages
+RUN apt-get update && \
+    apt-get -y install libdccl3-dev:armhf \
+            libcrypto++-dev:armhf \
+            libcrypto++6:armhf \
+            libwt-dev:armhf libwtdbo-dev:armhf libwtdbosqlite-dev:armhf libwthttp-dev:armhf \
+            libboost-regex-dev:armhf libicu-dev:armhf \
+            libhdf5-dev:armhf \
+            libgmp3-dev:armhf \
+            libstdc++-7-dev:armhf \
+            && \
+    rm -rf /var/lib/apt/lists/*

--- a/.circleci/docker/focal/base/Dockerfile
+++ b/.circleci/docker/focal/base/Dockerfile
@@ -1,0 +1,41 @@
+# creates gobysoft/goby3-ubuntu-build-base:20.04.N
+FROM gobysoft/ubuntu-build-base:20.04.1
+
+WORKDIR /root
+
+SHELL ["/bin/bash", "-c"]
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update && apt-get -y install dirmngr && \
+    echo "deb http://packages.gobysoft.org/ubuntu/release/ focal/" >> /etc/apt/sources.list.d/gobysoft_release.list && \
+    apt-key adv --recv-key --keyserver keyserver.ubuntu.com 19478082E2F8D3FE
+
+# Clone the Debian packaging project and install the dependencies it declares
+RUN git clone \
+    https://github.com/GobySoft/goby-debian -b 3.0 debian    
+
+RUN apt-get update && \
+    mk-build-deps -t "apt-get -y --no-install-recommends" -i "debian/control"
+
+# Comment out troublesome (not fully multi-arch compatible) dependencies
+RUN sed -i 's/^\( *\)clang/#\1clang/' debian/control && \
+    sed -i 's/^\(.*\)lld/#\1lld/' debian/control && \
+    sed -i 's/^\(.*\)lmodern/#\1lmodern/' debian/control && \
+    sed -i 's/^\(.*\)libwt/#\1libwt/' debian/control && \
+    sed -i 's/^\(.*\)libhdf/#\1libhdf/' debian/control && \
+    sed -i 's/^\(.*\)libboost-regex/#\1libboost-regex/' debian/control && \
+    sed -i 's/^\(.*\)libdccl/#\1libdccl/' debian/control && \
+    sed -i 's/^\(.*\)libgmp3/#\1libgmp3/' debian/control && \
+    sed -i 's/^\(.*\)libmavlink/#\1libmavlink/' debian/control && \
+    cat debian/control
+
+RUN apt-get update && \
+    ARCHS=(${CROSS_COMPILE_ARCHS}) && \
+    for ARCH in "${ARCHS[@]}"; \
+    do mk-build-deps -a ${ARCH} --host-arch ${ARCH} --build-arch amd64 -t "apt-get -y --no-install-recommends -o Debug::pkgProblemResolver=yes" -i "debian/control"; \
+    done && \
+    rm -rf /var/lib/apt/lists/*
+
+# Restore control file
+RUN git -C debian checkout control

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -47,9 +47,9 @@ endif()
 
 # If we're using clang, use lld as well
 if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
-  find_program(LDD_BINARY lld)
-  if(LDD_BINARY)
-    message(STATUS "Using Clang, setting to also use ldd")
+  find_program(LLD_BINARY lld)
+  if(LLD_BINARY)
+    message(STATUS "Using Clang, setting to also use lld")
     set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")
   else()
     message(STATUS "Using Clang, but lld is not installed, so using system default (probably ld)")    

--- a/scripts/goby_launch
+++ b/scripts/goby_launch
@@ -97,6 +97,9 @@ do_kill()
     done
 
     spid=()
+
+    rm $launchfile;
+    
     exit 0;
 }
 

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -39,8 +39,6 @@ endif()
 
 #proj
 goby_find_required_package(Proj)
-# goby and MOOS make use of deprecated PROJ4 API
-add_definitions(-DACCEPT_USE_OF_DEPRECATED_PROJ_API_H)
 
 #shared library suffix
 add_definitions(-DSHARED_LIBRARY_SUFFIX="${CMAKE_SHARED_LIBRARY_SUFFIX}")

--- a/src/acomms/modemdriver/benthos_atm900_driver_fsm.h
+++ b/src/acomms/modemdriver/benthos_atm900_driver_fsm.h
@@ -541,4 +541,6 @@ struct TransmitData : boost::statechart::state<TransmitData, Online>, StateNotif
 } // namespace acomms
 } // namespace goby
 
+#include "goby/acomms/modemdriver/detail/boost_statechart_compat.h"
+
 #endif

--- a/src/acomms/modemdriver/detail/boost_statechart_compat.h
+++ b/src/acomms/modemdriver/detail/boost_statechart_compat.h
@@ -1,0 +1,32 @@
+// Copyright 2009-2018 Toby Schneider (http://gobysoft.org/index.wt/people/toby)
+//                     GobySoft, LLC (2013-)
+//                     Massachusetts Institute of Technology (2007-2014)
+//                     Community contributors (see AUTHORS file)
+//
+//
+// This file is part of the Goby Underwater Autonomy Project Libraries
+// ("The Goby Libraries").
+//
+// The Goby Libraries are free software: you can redistribute them and/or modify
+// them under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2.1 of the License, or
+// (at your option) any later version.
+//
+// The Goby Libraries are distributed in the hope that they will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Goby.  If not, see <http://www.gnu.org/licenses/>.
+
+// fixes linker problem with Boost 1.71 and Clang 10 with -g enabled:
+// /usr/bin/ld: ../../../../lib/libgoby.so.3.0.0~beta10+11+g47dd90bc-dirty: undefined reference to `boost::statechart::detail::no_context<goby::acomms::benthos::fsm::EvReset>::no_function(goby::acomms::benthos::fsm::EvReset const&)'
+
+#if __clang_major__ == 10
+#ifndef GOBY_WORKAROUND_BOOST_STATECHART_NO_FUNCTION
+#define GOBY_WORKAROUND_BOOST_STATECHART_NO_FUNCTION
+template <class Event>
+void boost::statechart::detail::no_context<Event>::no_function(const Event&){};
+#endif
+#endif

--- a/src/acomms/modemdriver/iridium_driver_fsm.h
+++ b/src/acomms/modemdriver/iridium_driver_fsm.h
@@ -735,4 +735,6 @@ struct SBDReceive : boost::statechart::state<SBDReceive, SBD>, StateNotify
 } // namespace acomms
 } // namespace goby
 
+#include "goby/acomms/modemdriver/detail/boost_statechart_compat.h"
+
 #endif

--- a/src/apps/middleware/clang_tool/visualize.cpp
+++ b/src/apps/middleware/clang_tool/visualize.cpp
@@ -76,7 +76,7 @@ struct Application
                 std::set<std::string> bases;
                 if (bases_node)
                 {
-                    for (auto base : bases_node) bases.insert(base.as<std::string>());
+                    for (const auto& base : bases_node) bases.insert(base.as<std::string>());
                 }
 
                 threads.emplace(thread_name,
@@ -88,7 +88,7 @@ struct Application
             {
                 auto& bases = thread_p.second->bases;
                 bool is_direct_thread_subclass = false;
-                for (const auto base : bases)
+                for (const auto& base : bases)
                 {
                     if (base.find("goby::middleware::SimpleThread") == 0)
                         is_direct_thread_subclass = true;
@@ -98,7 +98,7 @@ struct Application
                 {
                     for (auto& base_thread_p : threads)
                     {
-                        for (const auto base : bases)
+                        for (const auto& base : bases)
                         {
                             if (base == base_thread_p.first)
                             {

--- a/src/goby-config.cmake.in
+++ b/src/goby-config.cmake.in
@@ -26,6 +26,16 @@ foreach(component ${GOBY_FIND_COMPONENTS})
   endif()
 endforeach()
 
+include(CMakeFindDependencyMacro)
+
+if(${CMAKE_VERSION} VERSION_GREATER_EQUAL 3.9)
+find_dependency(Boost 1.58.0 COMPONENTS system date_time program_options filesystem)
+find_dependency(DCCL)
+# Todo - install these Find scripts for use by downstream projects
+#find_dependency(Proj)
+#find_dependency(ProtobufGoby)
+endif()
+
 get_target_property(GOBY_INCLUDE_DIR goby INTERFACE_INCLUDE_DIRECTORIES)
 
 # goby_clang_tool interfaces files

--- a/src/middleware/log/hdf5/hdf5.cpp
+++ b/src/middleware/log/hdf5/hdf5.cpp
@@ -389,7 +389,7 @@ void goby::middleware::hdf5::Writer::write_vector(const std::string& group,
     std::vector<char> data_char;
     std::vector<hsize_t> hs = hs_outer;
 
-    int max_size = 0;
+    size_t max_size = 0;
     for (unsigned i = 0, n = data.size(); i < n; ++i)
     {
         if (data[i].size() > max_size)

--- a/src/moos/goby_moos_app.h
+++ b/src/moos/goby_moos_app.h
@@ -853,8 +853,7 @@ void goby::moos::GobyMOOSAppSelector<MOOSAppType>::read_configuration(
         fetch_moos_globals(cfg, moos_file_reader);
 
         // add / overwrite any options that are specified in the cfg file with those given on the command line
-        typedef std::pair<std::string, boost::program_options::variable_value> P;
-        for (const P& p : var_map)
+        for (const auto& p : var_map)
         {
             // let protobuf deal with the defaults
             if (!p.second.defaulted())

--- a/src/moos/moos_protobuf_helpers.h
+++ b/src/moos/moos_protobuf_helpers.h
@@ -1511,11 +1511,6 @@ template <> class MOOSTranslation<protobuf::TranslatorEntry::TECHNIQUE_FORMAT>
 } // namespace goby
 
 /// \file moos_protobuf_helpers.h Helpers for MOOS applications for serializing and parsed Google Protocol buffers messages
-
-/// \brief Converts the Google Protocol Buffers message `msg` into a suitable (human readable) string `out` for sending via MOOS
-///
-/// \param out pointer to std::string to store serialized result
-/// \param msg Google Protocol buffers message to serialize
 namespace goby
 {
 namespace moos
@@ -1545,6 +1540,10 @@ inline void set_moos_technique(const goby::moos::protobuf::GobyMOOSAppConfig& cf
 } // namespace moos
 } // namespace goby
 
+/// \brief Converts the Google Protocol Buffers message `msg` into a suitable (human readable) string `out` for sending via MOOS
+///
+/// \param out pointer to std::string to store serialized result
+/// \param msg Google Protocol buffers message to serialize
 inline bool serialize_for_moos(std::string* out, const google::protobuf::Message& msg)
 {
     switch (goby::moos::get_moos_technique())

--- a/src/moos/transitional/message_algorithms.cpp
+++ b/src/moos/transitional/message_algorithms.cpp
@@ -112,7 +112,7 @@ void goby::moos::transitional::DCCLAlgorithmPerformer::check_algorithm(const std
     for (std::vector<std::string>::size_type i = 1, n = ref_vars.size(); i < n; ++i)
     {
         bool ref_found = false;
-        for (const std::shared_ptr<DCCLMessageVar> mv : msg.header_const())
+        for (const std::shared_ptr<DCCLMessageVar>& mv : msg.header_const())
         {
             if (ref_vars[i] == mv->name())
             {
@@ -121,7 +121,7 @@ void goby::moos::transitional::DCCLAlgorithmPerformer::check_algorithm(const std
             }
         }
 
-        for (const std::shared_ptr<DCCLMessageVar> mv : msg.layout_const())
+        for (const std::shared_ptr<DCCLMessageVar>& mv : msg.layout_const())
         {
             if (ref_vars[i] == mv->name())
             {

--- a/src/test/middleware/log/test.cpp
+++ b/src/test/middleware/log/test.cpp
@@ -116,7 +116,7 @@ void read_log(int test)
         bool expected_eof = false;
         assert(expected_eof);
     }
-    catch (std::ifstream::failure e)
+    catch (std::ifstream::failure& e)
     {
         assert(in_log_file.eof());
     }

--- a/src/test/util/ais/ais.cpp
+++ b/src/test/util/ais/ais.cpp
@@ -19,8 +19,16 @@
 // You should have received a copy of the GNU General Public License
 // along with Goby.  If not, see <http://www.gnu.org/licenses/>.
 
+#include <boost/version.hpp>
+
 #define BOOST_TEST_MODULE ais_codec_test
+
+#if BOOST_VERSION > 105800
+#include <boost/test/tools/floating_point_comparison.hpp>
+#else
 #include <boost/test/floating_point_comparison.hpp>
+#endif
+
 #include <boost/test/included/unit_test.hpp>
 #include <boost/units/io.hpp>
 #include <boost/units/systems/si/prefixes.hpp>

--- a/src/util/ais/decode.h
+++ b/src/util/ais/decode.h
@@ -169,6 +169,7 @@ inline ostream& operator<<(ostream& os, Decoder::ParsedType t)
 {
     switch (t)
     {
+        default:
         case Decoder::ParsedType::NOT_SUPPORTED: return os << "NOT_SUPPORTED";
         case Decoder::ParsedType::VOYAGE: return os << "VOYAGE";
         case Decoder::ParsedType::POSITION: return os << "POSITION";

--- a/src/util/debug_logger/flex_ostreambuf.cpp
+++ b/src/util/debug_logger/flex_ostreambuf.cpp
@@ -109,8 +109,7 @@ void goby::util::FlexOStreamBuf::enable_gui()
     curses_->startup();
 
     // add any groups already on the screen as ncurses windows
-    typedef std::pair<std::string, logger::Group> P;
-    for (const P& p : groups_) curses_->add_win(&groups_[p.first]);
+    for (const auto& p : groups_) curses_->add_win(&groups_[p.first]);
 
     curses_->recalculate_win();
 

--- a/src/util/debug_logger/term_color.h
+++ b/src/util/debug_logger/term_color.h
@@ -175,8 +175,7 @@ class TermColor
     // red -> "red"
     std::string priv_str_from_col(const Colors::Color& c)
     {
-        typedef std::pair<std::string, Colors::Color> P;
-        for (const P& p : colors_map_)
+        for (const auto& p : colors_map_)
         {
             if (p.second == c)
                 return p.first;
@@ -190,8 +189,7 @@ class TermColor
     // red -> "\33[31m"
     std::string priv_esc_code_from_col(const Colors::Color& c)
     {
-        typedef std::pair<std::string, Colors::Color> P;
-        for (const P& p : esc_code_map_)
+        for (const auto& p : esc_code_map_)
         {
             if (p.second == c)
                 return p.first;

--- a/src/util/geodesy.h
+++ b/src/util/geodesy.h
@@ -29,6 +29,7 @@
 #include <boost/units/systems/angle/degrees.hpp>
 #include <boost/units/systems/si/length.hpp>
 
+#define ACCEPT_USE_OF_DEPRECATED_PROJ_API_H
 #include <proj_api.h>
 
 #include "goby/util/sci.h"

--- a/src/zeromq/transport/interprocess.cpp
+++ b/src/zeromq/transport/interprocess.cpp
@@ -27,6 +27,25 @@
 using goby::glog;
 using namespace goby::util::logger;
 
+// support moving to new API in ZeroMQ 4.3.1
+#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
+auto zmq_send_flags_none{zmq::send_flags::none};
+auto zmq_recv_flags_none{zmq::recv_flags::none};
+#else
+int zmq_send_flags_none{0};
+int zmq_recv_flags_none{0};
+#endif
+
+bool zmq_socket_recv(zmq::socket_t& socket, zmq::message_t& msg,
+                     goby::zeromq::zmq_recv_flags_type flags = zmq_recv_flags_none)
+{
+#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
+    return bool(socket.recv(msg, flags));
+#else
+    return socket.recv(&msg, flags);
+#endif
+}
+
 void goby::zeromq::setup_socket(zmq::socket_t& socket, const protobuf::Socket& cfg)
 {
     int send_hwm = cfg.send_queue_size();
@@ -67,11 +86,11 @@ goby::zeromq::InterProcessPortalMainThread::InterProcessPortalMainThread(zmq::co
 }
 
 bool goby::zeromq::InterProcessPortalMainThread::recv(protobuf::InprocControl* control_msg,
-                                                      int flags)
+                                                      zmq_recv_flags_type flags)
 {
     zmq::message_t zmq_msg;
     bool message_received = false;
-    if (control_socket_.recv(&zmq_msg, flags))
+    if (zmq_socket_recv(control_socket_, zmq_msg, flags))
     {
         control_msg->ParseFromArray((char*)zmq_msg.data(), zmq_msg.size());
         glog.is(DEBUG3) &&
@@ -104,7 +123,8 @@ void goby::zeromq::InterProcessPortalMainThread::publish(const std::string& iden
         zmq::message_t msg(identifier.size() + size);
         memcpy(msg.data(), identifier.data(), identifier.size());
         memcpy(static_cast<char*>(msg.data()) + identifier.size(), bytes, size);
-        publish_socket_.send(msg);
+
+        publish_socket_.send(msg, zmq_send_flags_none);
 
         glog.is(DEBUG3) && glog << "Published " << size << " bytes to ["
                                 << identifier.substr(0, identifier.size() - 1) << "]" << std::endl;
@@ -157,7 +177,8 @@ void goby::zeromq::InterProcessPortalMainThread::send_control_msg(
 {
     zmq::message_t zmq_control_msg(control.ByteSize());
     control.SerializeToArray((char*)zmq_control_msg.data(), zmq_control_msg.size());
-    control_socket_.send(zmq_control_msg);
+
+    control_socket_.send(zmq_control_msg, zmq_send_flags_none);
 }
 
 //
@@ -217,7 +238,7 @@ void goby::zeromq::InterProcessPortalReadThread::run()
 
             zmq::message_t msg(req.ByteSize());
             req.SerializeToArray(static_cast<char*>(msg.data()), req.ByteSize());
-            manager_socket_.send(msg);
+            manager_socket_.send(msg, zmq_send_flags_none);
 
             auto start = goby::time::SystemClock::now();
             while (!have_pubsub_sockets_ &&
@@ -245,15 +266,15 @@ void goby::zeromq::InterProcessPortalReadThread::poll(long timeout_ms)
             switch (i)
             {
                 case SOCKET_CONTROL:
-                    if (control_socket_.recv(&zmq_msg))
+                    if (zmq_socket_recv(control_socket_, zmq_msg))
                         control_data(zmq_msg);
                     break;
                 case SOCKET_SUBSCRIBE:
-                    if (subscribe_socket_.recv(&zmq_msg))
+                    if (zmq_socket_recv(subscribe_socket_, zmq_msg))
                         subscribe_data(zmq_msg);
                     break;
                 case SOCKET_MANAGER:
-                    if (manager_socket_.recv(&zmq_msg))
+                    if (zmq_socket_recv(manager_socket_, zmq_msg))
                         manager_data(zmq_msg);
                     break;
             }
@@ -341,7 +362,7 @@ void goby::zeromq::InterProcessPortalReadThread::send_control_msg(
 {
     zmq::message_t zmq_control_msg(control.ByteSize());
     control.SerializeToArray((char*)zmq_control_msg.data(), zmq_control_msg.size());
-    control_socket_.send(zmq_control_msg);
+    control_socket_.send(zmq_control_msg, zmq_send_flags_none);
     poller_cv_->notify_all();
 }
 
@@ -402,7 +423,11 @@ void goby::zeromq::Router::run()
     }
     try
     {
+#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
+        zmq::proxy(frontend, backend);
+#else
         zmq::proxy((void*)frontend, (void*)backend, nullptr);
+#endif
     }
     catch (const zmq::error_t& e)
     {
@@ -445,7 +470,7 @@ void goby::zeromq::Manager::run()
         while (true)
         {
             zmq::message_t request;
-            socket.recv(&request);
+            zmq_socket_recv(socket, request);
 
             protobuf::ManagerRequest pb_request;
             pb_request.ParseFromArray((char*)request.data(), request.size());
@@ -497,7 +522,7 @@ void goby::zeromq::Manager::run()
 
             zmq::message_t reply(pb_response.ByteSize());
             pb_response.SerializeToArray((char*)reply.data(), reply.size());
-            socket.send(reply);
+            socket.send(reply, zmq_send_flags_none);
         }
     }
     catch (const zmq::error_t& e)

--- a/src/zeromq/transport/interprocess.cpp
+++ b/src/zeromq/transport/interprocess.cpp
@@ -28,21 +28,21 @@ using goby::glog;
 using namespace goby::util::logger;
 
 // support moving to new API in ZeroMQ 4.3.1
-#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
-auto zmq_send_flags_none{zmq::send_flags::none};
-auto zmq_recv_flags_none{zmq::recv_flags::none};
-#else
+#ifdef USE_OLD_ZMQ_CPP_API
 int zmq_send_flags_none{0};
 int zmq_recv_flags_none{0};
+#else
+auto zmq_send_flags_none{zmq::send_flags::none};
+auto zmq_recv_flags_none{zmq::recv_flags::none};
 #endif
 
 bool zmq_socket_recv(zmq::socket_t& socket, zmq::message_t& msg,
                      goby::zeromq::zmq_recv_flags_type flags = zmq_recv_flags_none)
 {
-#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
-    return bool(socket.recv(msg, flags));
-#else
+#ifdef USE_OLD_ZMQ_CPP_API
     return socket.recv(&msg, flags);
+#else
+    return bool(socket.recv(msg, flags));
 #endif
 }
 
@@ -423,10 +423,10 @@ void goby::zeromq::Router::run()
     }
     try
     {
-#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
-        zmq::proxy(frontend, backend);
-#else
+#ifdef USE_OLD_ZMQ_CPP_API
         zmq::proxy((void*)frontend, (void*)backend, nullptr);
+#else
+        zmq::proxy(frontend, backend);
 #endif
     }
     catch (const zmq::error_t& e)

--- a/src/zeromq/transport/interprocess.h
+++ b/src/zeromq/transport/interprocess.h
@@ -30,18 +30,22 @@
 #include "goby/zeromq/protobuf/interprocess_config.pb.h"
 #include "goby/zeromq/protobuf/interprocess_zeromq.pb.h"
 
+#if ZMQ_VERSION <= ZMQ_MAKE_VERSION(4, 3, 1)
+#define USE_OLD_ZMQ_CPP_API
+#endif
+
 namespace goby
 {
 namespace zeromq
 {
 void setup_socket(zmq::socket_t& socket, const protobuf::Socket& cfg);
 
-#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
-using zmq_recv_flags_type = zmq::recv_flags;
-using zmq_send_flags_type = zmq::send_flags;
-#else
+#ifdef USE_OLD_ZMQ_CPP_API
 using zmq_recv_flags_type = int;
 using zmq_send_flags_type = int;
+#else
+using zmq_recv_flags_type = zmq::recv_flags;
+using zmq_send_flags_type = zmq::send_flags;
 #endif
 
 // run in the same thread as InterProcessPortal
@@ -279,10 +283,10 @@ class InterProcessPortal
         int items = 0;
         protobuf::InprocControl control_msg;
 
-#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
-        auto flags = zmq::recv_flags::dontwait;
-#else
+#ifdef USE_OLD_ZMQ_CPP_API
         int flags = ZMQ_NOBLOCK;
+#else
+        auto flags = zmq::recv_flags::dontwait;
 #endif
 
         while (zmq_main_.recv(&control_msg, flags))

--- a/src/zeromq/transport/interprocess.h
+++ b/src/zeromq/transport/interprocess.h
@@ -35,13 +35,24 @@ namespace goby
 namespace zeromq
 {
 void setup_socket(zmq::socket_t& socket, const protobuf::Socket& cfg);
+
+#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
+using zmq_recv_flags_type = zmq::recv_flags;
+using zmq_send_flags_type = zmq::send_flags;
+#else
+using zmq_recv_flags_type = int;
+using zmq_send_flags_type = int;
+#endif
+
 // run in the same thread as InterProcessPortal
 class InterProcessPortalMainThread
 {
   public:
     InterProcessPortalMainThread(zmq::context_t& context);
     bool ready() { return publish_socket_configured_; }
-    bool recv(protobuf::InprocControl* control_msg, int flags = 0);
+
+    bool recv(protobuf::InprocControl* control_msg,
+              zmq_recv_flags_type flags = zmq_recv_flags_type());
     void set_publish_cfg(const protobuf::Socket& cfg);
     void publish(const std::string& identifier, const char* bytes, int size);
     void subscribe(const std::string& identifier);
@@ -267,7 +278,14 @@ class InterProcessPortal
     {
         int items = 0;
         protobuf::InprocControl control_msg;
-        while (zmq_main_.recv(&control_msg, ZMQ_NOBLOCK))
+
+#if ZMQ_VERSION >= ZMQ_MAKE_VERSION(4, 3, 1)
+        auto flags = zmq::recv_flags::dontwait;
+#else
+        int flags = ZMQ_NOBLOCK;
+#endif
+
+        while (zmq_main_.recv(&control_msg, flags))
         {
             switch (control_msg.type())
             {

